zapcore/sampler.go

```diff
@@ -126,9 +178,35 @@ func (s *sampler) Check(ent Entry, ce *CheckedEntry) *CheckedEntry {
 	}
 
 	counter := s.counts.get(ent.Level, ent.Message)
-	n := counter.IncCheckReset(ent.Time, s.tick)
+	n, prevN := counter.IncCheckReset(ent.Time, s.tick)
 	if n > s.first && (n-s.first)%s.thereafter != 0 {
 		return ce
 	}
+	if s.reporting.enabled && prevN > 0 {
+		s.report(ent, prevN)
+	}
 	return s.Core.Check(ent, ce)
 }
+
+func (s *sampler) report(ent Entry, n uint64) {
+	sampleCount := n - s.first
+	sampleCount = sampleCount - sampleCount/s.thereafter
+	if sampleCount <= 0 {
+		return
+	}
+	entry := Entry{
+		LoggerName: s.reporting.loggerName,
+		Time:       time.Now(),
+		Level:      s.reporting.level,
+		Message:    s.reporting.message,
+	}
+	fields := []Field{
+		{Key: "original_level", Type: StringType, String: ent.Level.String()},
+		{Key: "original_message", Type: StringType, String: ent.Message},
+		{Key: "count", Type: Int64Type, Integer: int64(sampleCount)},
+	}
+	err := s.reporting.core.Write(entry, fields)
+	if err != nil {
+		fmt.Printf("%v write error: %v\n", time.Now(), err)
```

Comment:
This is what `Logger.errorOutput` is for, but probably this `report` method should just return an error, and be used in existing error handling paths... The fact that `With` above doesn't support error propagation, seems a sign that we shouldn't be writing output during a `With`, but instead defer reporting until actual attempt to log time.

Reply:
This one is called by `Check()` which doesn't support error propagation either. I thought printing would be better than just ignoring.